### ISSUE TITLE
Add a `like` argument to creation ops

### ISSUE THREAD

**Issue by n2cholas** [pytorch/pytorch#40419]
## 🚀 Feature
Add a `like` argument to all the creation ops such that, given a tensor `x`, `torch.eye(3, like=x)` has the same semantics as `torch.eye(3, dtype=x.dtype, device=x.device)`.

## Motivation

In many cases, you want to create a tensor with a matching `dtype` and `device` as some input to do computations with it. The idiomatic way to do so is as follows:

```python
def f(input):
    other = torch.eye(3, dtype=input.dtype, device=input.device)
    # do some computation with input and other
```

However, this is quite verbose. Very often, people write the following code instead:
```python
def f(input):
    other = torch.eye(3).to(input)
```
This works, but has a couple of drawbacks:

1. We must create a tensor then potentially copy it to a different device, whereas in the first code example the tensor was created on the right device. This uses unnecessary memory and time.
2. `.to()` by default is a blocking operation (e.g. for a host->device transfer), which can have significant performance implications for python-heavy code. `torch.eye(3, dtype=input.dtype, device=input.device)` does the creation asynchronously.

## Pitch

Adding a `like` argument to all the creation ops such that, given a tensor `x`, `torch.eye(3, like=x)` has the same semantics as `torch.eye(3, dtype=x.dtype, device=y.device)` will alleviate this issue. This syntax is as succinct and clear as using `.to()`, so people will not use the less efficient method for clarity or out of laziness.

The current API already has functions of the form `torch.*_like` (e.g. `torch.ones_like`), which has similar behaviour to what I described above (except it also copies the shape). Thus, this new argument has some API symmetry, making it easy for new users to learn and existing users to adopt.

## Alternatives

Our new `like` argument can also ensure the new tensor has the same layout as the original, but `.to()` currently does not do this, and I think it would be best if `like` follows the semantics of `.to()`.

Instead of adding an argument, we can just add corresponding `torch.*_like` functions for all the creation ops. However, I think this would be confusing: `torch.*_like` creates tensors of the same shape, dtype, and device as the input; however most of the creation ops do not support making a tensor of the same shape necessarily (e.g. `torch.eye`, `torch.arange`, `torch.linspace`, etc). Furthermore, the signature would be confusing as `torch.*_like` functions typically have the other tensor as the first argument.

Thanks for considering this request, and please let me know if you have any feedback!

**Comment by SsnL** [comment]
Sounds more like the `tensor.new_*` family. Those currently don't include things like `eye`.

**Comment by n2cholas** [comment]
Good point @SsnL! Adding `tensor.new_*` versions of the rest of the creation ops (e.g. `tensor.new_linspace`, `tensor.new_eye`, etc.) would be a valid and equally useful alternative to this proposal.

**Comment by vadimkantorov** [comment]
Related (though rejected) feature request for `eye_like` (with default `dim1=-1, dim2 = -2`): https://github.com/pytorch/pytorch/issues/9244

That aside, I also support `like` idea that would be similar to tensor options in C++? `new_` methods are okay for (few) factory methods, `like` could also be introduced to some other functions that are practically useful to specify extra dtype/device for output tensors

**Comment by zou3519** [comment]
I agree that it would be consistent if we put such an operator into the `tensor.new_*` family.

@vadimkantorov did you have specific operators in mind?

**Comment by vadimkantorov** [comment]
@zou3519 Two from my experience: https://github.com/pytorch/pytorch/issues/32570 https://github.com/pytorch/pytorch/issues/33046 where specifying output dtype is useful (not device though). This would mostly be cases of saving memory and not dealing with specifying out tensor manually.

**Comment by vadimkantorov** [comment]
This is also useful for random factory methods, e.g. we have rand_like, randn_like but no bernoulli_like, multinomial_like etc. 

Usecase for some bernoulli_like op: https://github.com/pytorch/pytorch/blob/61623430d32c081124067fc4abf33bd4a023af90/torch/csrc/jit/runtime/symbolic_script.cpp#L1096
